oval/redhat/redhat_test.go

```diff
@@ -142,5 +143,44 @@ func TestConfig_Update(t *testing.T) {
 			assert.NoError(t, err, tc.name)
 		})
 	}
+}
+
+func TestConfig_saveRHSAPerYear(t *testing.T) {
```

Comment:
I've already written tests for `Update`. At the same time, we can test the behavior of `saveRHSAPerYear` through `Update`. These test cases are already tested in `TestConfig_Update`. This method is private. Basically, I think we don't have to test private methods. It is sometimes difficult to test a private method via a public method. Or, we sometimes want to test many cases. In that case, I write unit tests even for a private method. This is a little famous religious war...

Reply:
My 2¢: I fully agree with testing exported methods first and I also believe in testing unexported methods through the means of calling exported methods that in turn call the former. 

But if calling an exported method will never invoke the unexported method in question under a test environment, then the unexported method should be removed. It is essentially dead code at that point that will never see the light of the day in real usage. Of course this is all based on the premise that we're writing tests as close as possible to the real environment. 

By definition, a unit of testing in a unit test is a method. If we're not testing each method, exported or not, we're not testing each unit independently.

Reply:
I completely agree with your idea that uncalled functions should be removed.

>If we're not testing each method, exported or not, we're not testing each unit independently.
So, does it mean that we should also test private methods based on unit test definitions?

A unit test is useful for refactoring. If there is a unit test, we can change the internal code without changing the behavior from the outside. But if we write a test for unexported functions, it sometimes prevents refactoring. This is because that test is for implementation, not for behavior. This is not an absolute rule. We should write a test for an unexported function sometimes.

Reply:
Agreed! On that point, if something's hard to unit test it should be refactored out. In this case it seems the without actually refactoring the unexported method, it will be hard to test through the exported one.

Reply:
Hmm... I didn't feel hard to test in this case because I just divided one function into the exported one and the unexported one. Which points do you feel hard? How should we improve?

Reply:
Without having a test for `saveRHSAPerYear` this codepath is untested: https://github.com/aquasecurity/vuln-list-update/pull/19/files#diff-ceaa4154ecb6fbceb0b0479e253bbdb6R99-R102

Reply:
I understand. I've already passed the XML including an invalid RHSA-ID in TestConfig_Update. So this codepath (https://github.com/aquasecurity/vuln-list-update/commit/b8e8adf4966f1c05a2f839f9fefc88f3c81b1b62#diff-ceaa4154ecb6fbceb0b0479e253bbdb6R103-R106) is already tested, but that codepath was not tested. Although I knew it, I thought it was too much. In this case, as you said, it might have been better to test it. Thanks.